src/NamingStrategy/UnderscoreNamingStrategy.php

```diff
@@ -42,39 +45,108 @@ public function hydrate($name)
      */
     public function extract($name)
     {
-        return $this->getCamelCaseToUnderscoreFilter()->filter($name);
+        $filter = $this->getCamelCaseToUnderscoreFilter();
+
+        return $filter($name);
     }
 
     /**
-     * @return FilterChain
+     * @return Closure
      */
     protected function getUnderscoreToStudlyCaseFilter()
     {
-        if (static::$underscoreToStudlyCaseFilter instanceof FilterChain) {
+        if (static::$underscoreToStudlyCaseFilter instanceof Closure) {
             return static::$underscoreToStudlyCaseFilter;
         }
 
-        $filter = new FilterChain();
+        return static::$underscoreToStudlyCaseFilter = function ($value) {
```

Comment:
Code should also probably be extracted to an external class to avoid going completely crazy when testing it.